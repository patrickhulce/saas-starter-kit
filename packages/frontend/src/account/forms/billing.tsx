import * as React from 'react'
import {IUser} from '../../../../shared/lib/typedefs'

export class BillingForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2>Billing</h2>,
    ]
  }
}
