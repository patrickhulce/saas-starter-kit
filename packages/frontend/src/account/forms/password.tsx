import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'

export class PasswordForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2 key="password">Password</h2>,
    ]
  }
}
