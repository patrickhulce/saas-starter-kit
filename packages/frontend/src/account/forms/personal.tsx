import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'
import {ErrorBar} from '../../components/error-bar/error-bar'
import {Form, IFormData} from '../../components/form'
import {LoadingBar} from '../../components/loading-bar/loading-bar'

async function updateAccount(user: IUser): Promise<void> {
  const response = await fetch(`/api/v1/users/${user.id}`, {
    method: 'PUT',
    body: JSON.stringify(user),
    headers: {'content-type': 'application/json'},
  })

  if (response.status !== 200) {
    // TODO: Provide more specific error message
    throw new Error('Error updating account')
  }
}

interface INamesFormState {
  isLoading?: boolean
  errorMessage?: string
}

class NamesForm extends Form<{user: IUser}, INamesFormState> {
  public state: INamesFormState = {}

  protected async _handleSubmit(data: IFormData): Promise<void> {
    const user: IUser = {...this.props.user, firstName: data.first, lastName: data.last}
    await updateAccount(user)
  }

  public render(): JSX.Element {
    return (
      <form onSubmit={this.handleSubmit}>
        <LoadingBar isLoading={this.state.isLoading} />
        <ErrorBar message={this.state.errorMessage} />
        <TextField
          id="first-name"
          name="first"
          label="First Name"
          defaultValue={this.props.user.firstName}
          required
        />
        <TextField
          id="last-name"
          name="last"
          label="Last Name"
          defaultValue={this.props.user.lastName}
          required
        />
        <Button variant="raised" color="primary" type="submit">
          Update
        </Button>
      </form>
    )
  }
}

export class PersonalForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2 key="header">Personal Information</h2>,
      <NamesForm key="names-form" user={this.props.user} />,
    ]
  }
}
