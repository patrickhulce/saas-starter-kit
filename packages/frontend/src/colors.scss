$primary-brand: #0771b8;
$accent-brand: #fcb827;

$darker-gray: #333;
$dark-gray: #444;
$medium-gray: #666;
$lightest-gray: #eee;

$dark-red: #721c24;
$light-pink: #f5c6c6;
$lightest-pink: #fff4f4;

$dark-green: #1c721c;
$light-green: #c6f5cd;
$lightest-green: #f5fff4;
