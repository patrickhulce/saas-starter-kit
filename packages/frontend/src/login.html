<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>THE_PRODUCT_DISPLAY_NAME Login</title>
</head>

<body>
  <h1 class="app-name">THE_PRODUCT_DISPLAY_NAME</h1>
  <div id="app-root"></div>
</body>

</html>
