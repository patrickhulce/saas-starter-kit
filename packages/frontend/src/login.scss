:global {
  html {
    height: 100%;
  }

  body {
    background: #0771b8;
    color: rgba(0, 0, 0, 0.87);
    font-family: Roboto, Helvetica, Arial, sans-serif;
    height: 100%;
    margin: 0;
    padding: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
  }

  h1.app-name {
    color: white;
    font-size: 48px;
    font-weight: bold;
    text-shadow: 0 1px 5px #444;
    margin-bottom: 40px;
  }

  #firebaseui-auth-container {
    padding: 15px;
    border-radius: 5px;
    background: white;
    box-shadow: 0 5px 15px #444;
    display: none;
  }

  #loader-container {
    height: 250px;
  }

  .loader,
  .loader:before,
  .loader:after {
    background: #ffffff;
    -webkit-animation: load1 1s infinite ease-in-out;
    animation: load1 1s infinite ease-in-out;
    width: 1em;
    height: 4em;
  }

  .loader {
    color: #ffffff;
    text-indent: -9999em;
    margin: 88px auto;
    position: relative;
    font-size: 11px;
    -webkit-transform: translateZ(0);
    -ms-transform: translateZ(0);
    transform: translateZ(0);
    -webkit-animation-delay: -0.16s;
    animation-delay: -0.16s;
  }

  .loader:before,
  .loader:after {
    position: absolute;
    top: 0;
    content: '';
  }

  .loader:before {
    left: -1.5em;
    -webkit-animation-delay: -0.32s;
    animation-delay: -0.32s;
  }

  .loader:after {
    left: 1.5em;
  }

  @-webkit-keyframes load1 {
    0%,
    80%,
    100% {
      box-shadow: 0 0;
      height: 4em;
    }
    40% {
      box-shadow: 0 -2em;
      height: 5em;
    }
  }

  @keyframes load1 {
    0%,
    80%,
    100% {
      box-shadow: 0 0;
      height: 4em;
    }
    40% {
      box-shadow: 0 -2em;
      height: 5em;
    }
  }
}
