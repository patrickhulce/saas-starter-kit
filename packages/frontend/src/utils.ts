import * as React from 'react'
import * as ReactDOM from 'react-dom'

import {IUser} from '../../shared/lib/typedefs'

import {getLoggedInUser, refreshLoggedInUser} from './services/user-service'

export async function redirectToLogin(): Promise<void> {
  window.location.href = '/login'
  return new Promise(resolve => setTimeout(resolve, 10000)) as Promise<void>
}

export async function redirectIfNotAuthorized(): Promise<void> {
  try {
    const localUser = await getLoggedInUser()
    if (!localUser) await redirectToLogin()
    await refreshLoggedInUser()
    const remoteUser = await getLoggedInUser()
    if (!remoteUser || localUser!.id !== remoteUser.id) await redirectToLogin()
  } catch (err) {
    await redirectToLogin()
  }
}

export async function findUserOnStartupOrBail(): Promise<IUser> {
  const localUser = await getLoggedInUser()
  if (!localUser) await redirectToLogin()
  setTimeout(redirectIfNotAuthorized, 0)
  return localUser!
}

export function createRenderFn(render: () => React.ReactElement<any>): () => void {
  return () => ReactDOM.render(render(), document.getElementById('app-root'))
}

export function HMR(nodeModule: any, handle: (nodeModule: any) => any): void {
  if (nodeModule.hot) {
    handle(nodeModule)
  }
}

export const testIds = {
  loadingBar: 'loading-bar',
  messageBar: 'message-bar',
  // Login IDs
  createAccountTab: 'create-account-tab',
  createAccountSubmit: 'create-account-submit',
  loginFormPrimaryBtn: 'login-form-primary-btn',
  loginFormSecondaryBtn: 'login-form-secondary-btn',
  loginForm: 'login-form',
  registerForm: 'register-form',
  profileNamesForm: 'profile-names-form',
}
