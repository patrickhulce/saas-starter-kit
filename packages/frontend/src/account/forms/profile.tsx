import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'
import {BasicTextField} from '../../components/basic-text-field'
import {Form, IFormData, IFormState} from '../../components/form/form'
import * as formStyles from '../../components/form/form.scss'
import {updateAccount} from '../../services/user-service'
import {testIds} from '../../utils'

class NamesForm extends Form<{user: IUser}> {
  public state: IFormState = {}
  public testId: string = testIds.profileNamesForm

  protected async _handleSubmit(data: IFormData): Promise<void> {
    await updateAccount(this.props.user.id, {firstName: data.firstName, lastName: data.lastName})
  }

  public renderInputUI(): JSX.Element {
    return (
      <React.Fragment>
        <BasicTextField name="firstName" defaultValue={this.props.user.firstName} />
        <BasicTextField name="lastName" defaultValue={this.props.user.lastName} />
      </React.Fragment>
    )
  }
}

export class ProfileForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element {
    const formClasses = `${formStyles.columnarForm} ${formStyles.fixedWidth}`
    return (
      <React.Fragment>
        <h2>Personal Information</h2>
        <NamesForm user={this.props.user} className={formClasses} />
      </React.Fragment>
    )
  }
}
