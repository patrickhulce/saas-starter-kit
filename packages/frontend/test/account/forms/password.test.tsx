import * as React from 'react'
import {fireEvent, render, wait} from 'react-testing-library'

import {IUser} from '../../../../shared/lib/typedefs'
import {PasswordForm} from '../../../src/account/forms/password'
import {testIds} from '../../../src/utils'
import {createFetchPromise} from '../../utils'

describe('account/forms/profile.tsx', () => {
  let fetchMock: jest.Mock
  let user: IUser

  beforeEach(() => {
    fetchMock = jest.fn()
    user = {id: 1, email: 'klay@example.com'} as IUser
    self.fetch = fetchMock
  })

  it('should render form', () => {
    const form = render(<PasswordForm user={user} />)
    expect(form.container).toMatchSnapshot()
  })

  it('should send update names request to server', async () => {
    const {getByTestId} = render(<PasswordForm user={user} />)
    const mockFetch = createFetchPromise()
    fetchMock.mockImplementation(mockFetch.fn)

    fireEvent.submit(getByTestId(testIds.defaultForm))
    await wait(() => getByTestId(testIds.loadingBar))

    expect(fetchMock).toHaveBeenCalled()
    expect(fetchMock.mock.calls[0]).toMatchSnapshot()
  })
})
