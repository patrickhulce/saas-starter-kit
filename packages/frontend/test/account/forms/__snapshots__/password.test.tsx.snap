// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account/forms/profile.tsx should render form 1`] = `
<div>
  <h2>
    Password
  </h2>
  <form
    class="columnarForm fixedWidth"
    data-testid="default-form"
  >
    <button
      class="MuiButtonBase-root-23 MuiButton-root-1 MuiButton-contained-10 MuiButton-containedPrimary-11 MuiButton-raised-13 MuiButton-raisedPrimary-14"
      data-testid="default-submit-btn"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label-2"
      >
        Request Password Reset
      </span>
      <span
        class="MuiTouchRipple-root-26"
      />
    </button>
  </form>
</div>
`;

exports[`account/forms/profile.tsx should send update names request to server 1`] = `
Array [
  "/api/v1/users/password-reset",
  Object {
    "body": "{\\"email\\":\\"klay@example.com\\"}",
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;
